desertbot/modules/commands/OutOfContext.py

```diff
@@ -0,0 +1,244 @@
+from twisted.plugin import IPlugin
+from desertbot.moduleinterface import IModule
+from desertbot.modules.commandinterface import BotCommand
+from desertbot.message import IRCMessage, TargetTypes
+from desertbot.response import IRCResponse, ResponseType
+from desertbot.utils import string
+from zope.interface import implementer
+from collections import OrderedDict
+from typing import List, Union
+from weakref import WeakKeyDictionary
+import random, time
+
+try:
+    import re2
+except ImportError:
+    import re as re2
+
+
+@implementer(IPlugin, IModule)
+class OutOfContext(BotCommand):
+    messageLimit = 50
+
+    def actions(self):
+        return super(OutOfContext, self).actions() + [("message-channel", 1, self.storeMessage),
+                                                      ("action-channel", 1, self.storeMessage)]
+
+    def triggers(self):
+        return ["outofcontext"]
+
+    def _add(self, message: IRCMessage):
+        """add - Adds a quote to the OutOfContext log. The quote will be pulled from a message line buffer."""
+        if len(message.parameterList) < 2:
+            return IRCResponse(ResponseType.Say, "Add what?", message.replyTo)
+        if message.targetType == TargetTypes.USER:
+            return IRCResponse(ResponseType.Say, "You can only add messages from channels.", message.replyTo)
+
+        regex = re2.compile(re2.escape(" ".join(message.parameterList[1:])), re2.IGNORECASE)
+        if len(self.messageStore) == 0 or message.channel not in self.messageStore:
+            return IRCResponse(ResponseType.Say, "Sorry, there are no messages in my buffer.", message.replyTo)
+
+        matches = list(filter(regex.search, self.messageStore[message.channel]))
+        if len(matches) == 0:
+            return IRCResponse(ResponseType.Say,
+                               "Sorry, that didn't match anything in my message buffer.",
+                               message.replyTo)
+        if len(matches) > 1:
+            return IRCResponse(ResponseType.Say,
+                               "Sorry, that matches too many lines in my message buffer.",
+                               message.replyTo)
+
+        todayDate = time.strftime("[%Y-%m-%d] [%H:%M]")
+        quote = f"{todayDate} {matches[0]}"
+        if message.replyTo not in self.storage:
+            self.storage[message.replyTo] = []
+        if quote.lower() in [x.lower() for x in self.storage[message.replyTo]]:
```

Comment:
Worth keeping in mind that this operation takes O(n) time, meaning it scales according to the size of the list. (At least, for python stdlib lists. Might be worth it to look into using set() instead, but I am unsure how this would complicate loading/saving to json.

Reply:
I got rid of it altogether. It existed because in the past you couldn't delete messages if they were the same but that's no longer the case because removebyid is a thing.